MIL Component - MILModel

myron opened the pull request:
Fix #3163 

A wrapper around backbone CNN classification model suitable for MIL

drbeh reviewed (changes requested):
Hi @myron, all your git commits need to be signed for MONAI. You should use `-s` in ALL your commits:
```
git commit -sm "Add commit message"
```

https://github.com/Project-MONAI/MONAI/pull/3236/checks?check_run_id=4074346048

myron commented:
> Hi @myron, all your git commits need to be signed for MONAI. You should use `-s` in ALL your commits:
> 
> ```
> git commit -sm "Add commit message"
> ```
> 
> https://github.com/Project-MONAI/MONAI/pull/3236/checks?check_run_id=4074346048

even for draft PRs ?  I thought this is just to have an initial discussion

myron commented:
Ok, signed and pushed

drbeh commented:
Thank you @myron. The draft PR is the work in progress to let the repo maintainer know what is being developed. So you can continue to improve it until it is ready to review (and passing all the tests)

I will try to give my comments earlier but please work on addressing the tests that are not passing. You can start from the easy one like [code formatting](https://github.com/Project-MONAI/MONAI/runs/4074518260?check_suite_focus=true). Please let me know if you have any question.

Nic-Ma commented:
Hi @myron ,

Please add guys in `reviewers` instead of `assignees`.

Thanks.

wyli commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,184 @@
+        else:
+            raise ValueError("Unsupported mil_mode: " + str(mil_mode))
+
+        self.myfc = nn.Linear(nfc, num_classes)
```
is it possible to use NetAdapter to implement the `mean` and `max` modes? https://github.com/Project-MONAI/MONAI/blob/3960a514a5f44a10aceab463e926a19bc0b8c79a/monai/networks/nets/netadapter.py#L20-L24?

myron replied:
that seems unnecessary. it removes 2 layers (and here we remove 1), and will make code more complicated

wyli commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,184 @@
+                    return hook
+
+                self.fhooks = []
+                net.layer1.register_forward_hook(forward_hook("layer1"))
```
This may be replaced by the torchvision's feature extractor module https://pytorch.org/blog/FX-feature-extraction-torchvision/ could you please look into it?

myron replied:
yes, that new functionality we discussed might be used to replace it. but this code is working, changing will require testing, we should finish everything first

drbeh commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,184 @@
+            if mil_mode not in ["mean", "max", "att", "att_trans", "att_trans2"]:
+                raise ValueError("Custom backbone is not supported for the mode:" + str(mil_mode))
+
+        if self.mil_mode in ["mean", "max"]:
+            pass
```
This lines can be removed.

myron replied:
If I remove it , it will raise a ValueError  at the end of the if/else statement

drbeh commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,184 @@
+
+        else:
+            # use a custom backbone (untested)
+            net = backbone
```
It would be nice that this class can accept the name of the torchvision model as well., like `"densenet161"`.
So it can check if the backbone is `str`, it imports the model like this:
https://github.com/Project-MONAI/MONAI/blob/3960a514a5f44a10aceab463e926a19bc0b8c79a/monai/networks/nets/torchvision_fc.py#L61

myron replied:
hmm, ok

myron replied:
but we also need to supply the dimension of features space, as well as we need to remove the last FC layer, it'll be hard to do an arbitrary torchvision model based on string alone

drbeh replied:
Right, but you are already accepting any arbitrary model that is derived from `nn.Module`. Also how is it different than `net = models.resnet50(pretrained=pretrained)`?

myron replied:
if someone wants to provide ResNet50 as an arbitrary model , they will have to do these steps:
net = models.resnet50(pretrained=pretrained) #init
net.fc = nn.Identity() #remove final FC
nfc = 2048 # define output features size

model = MILModel(backbone=net, backbone_nfeatures=nfc)

the final FC layer for ResNet50 is called .fc , but for other networks it can be called different, so for arbitrary network we don't know how to remove it

drbeh replied:
@myron, I see what you are saying but it is not very different than `netadapter`. The FC layer does not need to be replaces by identity, it can be removed like here: https://github.com/Project-MONAI/MONAI/blob/3960a514a5f44a10aceab463e926a19bc0b8c79a/monai/networks/nets/netadapter.py#L75

And this can work for a large family of torchvision models.

myron replied:
this solution is not generic. the final layer of torchvision model may or may not be FC layer , it depends on when FC layer was initialised. Furthermore, the final layer may be nn.Sequential which will also produce incorrect results.  Using this approach will hide many potential errors , I would avoid using it

drbeh commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,184 @@
+            net.fc = torch.nn.Identity()  # remove final linear layer
+
+            self.extra_outputs = {}
+            if mil_mode == "att_trans_pyramid":
+                # register hooks to capture outputs of intermediate layers
+                def forward_hook(layer_name):
+                    def hook(module, input, output):
+                        self.extra_outputs[layer_name] = output
+
+                    return hook
+
+                self.fhooks = []
+                net.layer1.register_forward_hook(forward_hook("layer1"))
+                net.layer2.register_forward_hook(forward_hook("layer2"))
+                net.layer3.register_forward_hook(forward_hook("layer3"))
+                net.layer4.register_forward_hook(forward_hook("layer4"))
```
This block of code should be outside of `if backbone is None:`.

myron replied:
no, this is specific for ResNet50,  it won't work for other backbones

myron replied:
In fact this mode for deeper pyramid integration currently is implemented only for ResNet50,  there is no generic way to do it for arbitrary network without manual input

drbeh commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,184 @@
+
+        self.mil_mode = mil_mode
+        print("MILModel with mode", mil_mode, "num_classes", num_classes)
+        n_trans = 4
```
Does this number depends on the model being used?
If yes, shouldn't be derived from the model?
If no, can it be provided as an input argument?

myron replied:
it doesn't depend on the model.  yes we can make it an argument

myron replied:
I made it an argument

drbeh commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,184 @@
+        self.mil_mode = mil_mode
+        print("MILModel with mode", mil_mode, "num_classes", num_classes)
+        n_trans = 4
+        trans_dropout = 0.0
```
Does it have to be always 0.0?

myron replied:
I used 0, but I suppose we can make it an argument

myron replied:
it's an argument now

myron commented:
I've updated this model, added enum and moved some options into arguments 
@drbeh  does it look better?

drbeh commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,206 @@
+            net.fc = torch.nn.Identity()  # remove final linear layer
+
+            self.extra_outputs = {}
+            if self.mil_mode == MilMode.ATT_TRANS_PYRAMID:
```
Are you going to have ` MilMode.ATT_TRANS_PYRAMID` only when the backbone is None?

myron replied:
yes, that mode we can only support for backbone None,  it requires knowledge of the model structure and it's difficult to support it for an arbitrary model. In fact, I'm not sure if we need to support arbitrary model at all

drbeh replied:
If `MILModel` is not going to work with arbitrary models (which makes sense), can you specify with which models it can work? 
Thanks

myron replied:
Currently, it works with any backbone (if someone removes FC layer manually, and provides number of output features),
The exception will be  MilMode.ATT_TRANS_PYRAMID which doesn't support arbitrary models at all 

The alternative is we don't support Arbitrary models at all

myron commented:
@drbeh 

I've updated the code based on our conversation and added unit tests


- I've added more parameters as options in init
- I've made it accept a string for the backbone to initialize a torchvision model based on string. But it will only work for torchvision models that have .fc layer defined (all resnets, etc). Some other models will raise an exception, and ask users to initialize manually
- I've added unit tests for various scenarios (they all passed OK). Torchscript also passed

Please review

drbeh commented:
> * I've added unit tests for various scenarios (they all passed OK). Torchscript also passed

Thank you @myron. Many of the checks are failing for this PR. Can you please check and fix them?

You can start with the formatting and type checking errors: https://github.com/Project-MONAI/MONAI/runs/4132608039?check_suite_focus=true#step:7:118

drbeh reviewed (commented):
@myron, one of the common error that I see is that you are using `torchvision` and importing it directly, so if it is not installed it will fail:
https://github.com/Project-MONAI/MONAI/runs/4132607911?check_suite_focus=true#step:9:33

Remember that the minimum dependency of MONAI is only `torch` and `numpy` so any other external library that you are importing should be either imported using `optionanl_import` or handled in other ways.

https://github.com/Project-MONAI/MONAI/blob/8c0ff376587797aabd2abbdb569fb691880fc094/monai/networks/nets/torchvision_fc.py#L17

drbeh commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,222 @@
+from enum import Enum
+from typing import Callable, Dict, List, Optional, Sequence, Union
```
Please remove unused imports.

myron replied:
done

drbeh commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,222 @@
+
+import torch
+import torch.nn as nn
+import torchvision
```
Please import `torchvision` as optional like here:
https://github.com/Project-MONAI/MONAI/blob/8c0ff376587797aabd2abbdb569fb691880fc094/monai/networks/nets/torchvision_fc.py#L17

myron replied:
done

drbeh replied:
The same things apply to the tests. Please change 
```python
from torchvision.models import densenet121
```
thanks.

drbeh commented:
Hi @myron, two points:
1. Please do not force push your commits. It will remove the the commit history and make the comments' references broken. If you pull first and then apply your changes, you shouldn't be in a need to force push. Usually, there are softer ways to resolve the git conflicts.
2. Please address all the failed checks before requesting for review, and let me know if you encountered any issue that was not clear to you or need help with.

Thanks

drbeh commented on tests/test_milmodel.py:
```diff
@@ -0,0 +1,91 @@
+TEST_CASE_MILMODEL.append([{"num_classes": 5, "backbone": "resnet18", "pretrained": True}, (2, 2, 3, 512, 512), (2, 5)])
+
+# custom backbone
+from torchvision.models import densenet121
```
All the imports should be on the top, and please use optional import.

myron replied:
ok

myron commented:
> Hi @myron, two points:
> 
>     1. Please do not force push your commits. It will remove the the commit history and make the comments' references broken. If you pull first and then apply your changes, you shouldn't be in a need to force push. Usually, there are softer ways to resolve the git conflicts.
> 
>     2. Please address all the failed checks before requesting for review, and let me know if you encountered any issue that was not clear to you or need help with.
> 
> 
> Thanks
@drbeh 
The DCO check - error here (in the list of checks) tells me to amend commit , and do  git push --force-with-lease
is there another way to update the code without creating a different PR ?

Anyway, I've fixed the mypy, flake8 typing errors, it should pass the check now

myron commented:
@Nic-Ma @drbeh 

I'm not sure why it doesn't pass the test. I see in error logs that it can't import torchvision
But I use optional import , so it should be working ??  can you help

```
======================================================================
ERROR: test_milmodel (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_milmodel
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\unittest\loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
  File "D:\a\MONAI\MONAI\tests\test_milmodel.py", line 54, in <module>
    backbone = models.densenet121(pretrained=False)
  File "D:\a\MONAI\MONAI\monai\utils\module.py", line 341, in __getattr__
    raise self._exception
  File "D:\a\MONAI\MONAI\monai\utils\module.py", line 300, in optional_import
    pkg = __import__(module)  # top level module
monai.utils.module.OptionalImportError: import torchvision.models (No module named 'torchvision').
```

Nic-Ma commented:
Hi @myron ,

Please add `test_milmodel` to below file to skip the min tests:
https://github.com/Project-MONAI/MONAI/blob/dev/tests/min_tests.py

Thanks.

drbeh commented:
> The DCO check - error here (in the list of checks) tells me to amend commit , and do  git push --force-with-lease
> is there another way to update the code without creating a different PR ?
@myron 
Are you using `git commit --signoff -m"Your message"` or simply `git commit -sm"Your message"` for all the commits that you are locally making?

myron commented:
> > The DCO check - error here (in the list of checks) tells me to amend commit , and do  git push --force-with-lease
> > is there another way to update the code without creating a different PR ?
> > @myron
> > Are you using `git commit --signoff -m"Your message"` or simply `git commit -sm"Your message"` for all the commits that you are locally making?

For these PR updates I use

```
git add -A
git commit --amend --no-edit --signoff
git rebase dev
git push --force-with-lease origin milmodel
```

myron commented:
> Hi @myron ,
> 
> Please add `test_milmodel` to below file to skip the min tests: https://github.com/Project-MONAI/MONAI/blob/dev/tests/min_tests.py
> 
> Thanks.

done

drbeh commented:
> For these PR updates I use
> 
> ```
> git add -A
> git commit --amend --no-edit --signoff
> git rebase dev
> git push --force-with-lease origin milmodel
> ```

@myron,  you need to do this only once. I think from now on you should be able to do this:
```
git pull
git add path_to_the_updated_file.py
git commit -sm"Update this file"
git push
```
or if you want to add all the changes:
```
git pull
git commit -sam"Update many files"
git push
```

Can you check if this works for you?

myron commented:
> > For these PR updates I use
> > ```
> > git add -A
> > git commit --amend --no-edit --signoff
> > git rebase dev
> > git push --force-with-lease origin milmodel
> > ```
> 
> @myron, you need to do this only once. I think from now on you should be able to do this:
> 
> ```
> git pull
> git add path_to_the_updated_file.py
> git commit -sm"Update this file"
> git push
> ```
> 
> or if you want to add all the changes:
> 
> ```
> git pull
> git commit -sam"Update many files"
> git push
> ```
> 
> Can you check if this works for you?

yes it worked fine!

I still see DCO — DCO error (incorrect signoff) , I followed it's instructions buy I still get this check with error

drbeh commented:
> 
> yes it worked fine!
> 
> I still see DCO — DCO error (incorrect signoff) , I followed it's instructions buy I still get this check with error

@myron I believe it is caused by the misconfiguration of your `git config`. Look at this commit that you made:
```
commit d8ba36e7bd501854768ceaeacaa2991f0effbe43
Author: am <am>
Date:   Mon Nov 1 16:51:11 2021 -0700

    [DLMED] MILmodel  PR
    
    Signed-off-by: myron <amyronenko@nvidia.com>
```

You can see that Author has invalid email address `Author: am <am>`. So this should fix your problem:
```
git config user.email "amyronenko@nvidia.com"
```

myron commented:
another check error I see is this to JIT.
but my function looks correct, and does not take a variable number of arguments. Any suggestions?

```
torch.jit.frontend.NotSupportedError: Compiled functions can't take variable number of arguments or use keyword-only arguments with defaults:
  File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 190
def _forward_unimplemented(self, *input: Any) -> None:
                                 ~~~~~~ <--- HERE
    r"""Defines the computation performed at every call.
'MILModel.calc_head' is being compiled since it was called from 'MILModel.forward'
  File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/monai/networks/nets/milmodel.py", line 224
    
        if not no_head:
            x = self.calc_head(x)
            ~~~~~~~~~~~~~~~~~~~~ <--- HERE
    
        return x
```

wyli commented:
> I had it already set all this time, except for the very push (long time ago)
> 
> > git config user.email
> > [amyronenko@nvidia.com](mailto:amyronenko@nvidia.com)
> 
> in .git/config
> 
> ```
> [user]
>         email = amyronenko@nvidia.com
>         name = myron
> ```
> 
> but every time I do
> 
> ```
> git commit --amend --no-edit --signoff
> or 
> git commit --ams
> ```
> 
> it still doesn't help

Perhaps you should add the email to your GitHub account https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-github-user-account/managing-email-preferences/setting-your-commit-email-address#setting-your-commit-email-address-on-github (there's an option to hide the actual address)

wyli commented:
> another check error I see is this to JIT. but my function looks correct, and does not take a variable number of arguments. Any suggestions?
> 
> ```
> torch.jit.frontend.NotSupportedError: Compiled functions can't take variable number of arguments or use keyword-only arguments with defaults:
>   File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 190
> def _forward_unimplemented(self, *input: Any) -> None:
>                                  ~~~~~~ <--- HERE
>     r"""Defines the computation performed at every call.
> 'MILModel.calc_head' is being compiled since it was called from 'MILModel.forward'
>   File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/monai/networks/nets/milmodel.py", line 224
>     
>         if not no_head:
>             x = self.calc_head(x)
>             ~~~~~~~~~~~~~~~~~~~~ <--- HERE
>     
>         return x
> ```

@ericspod could you please help here?

myron commented:
> > another check error I see is this to JIT. but my function looks correct, and does not take a variable number of arguments. Any suggestions?
> > ```
> > torch.jit.frontend.NotSupportedError: Compiled functions can't take variable number of arguments or use keyword-only arguments with defaults:
> >   File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/torch/nn/modules/module.py", line 190
> > def _forward_unimplemented(self, *input: Any) -> None:
> >                                  ~~~~~~ <--- HERE
> >     r"""Defines the computation performed at every call.
> > 'MILModel.calc_head' is being compiled since it was called from 'MILModel.forward'
> >   File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/monai/networks/nets/milmodel.py", line 224
> >     
> >         if not no_head:
> >             x = self.calc_head(x)
> >             ~~~~~~~~~~~~~~~~~~~~ <--- HERE
> >     
> >         return x
> > ```
> 
> @ericspod could you please help here?

I have figured it out. no need to help

myron commented:
> 's an option to hi

Yes, my email is already added to github all this time. and not hidden.
still not working :(

myron commented:
I ran into another JIT issue, it fails the test on the older Pytorch version (1.7, over a year old), it looks like at that time Enum wasn't supported in JIT. 

@wyli @Nic-Ma any suggestions?
we can remove Enum, and use strings for MilMode, or we can skip this test ?

thanks

```
RuntimeError:
Module 'MILModel' has no attribute 'mil_mode' (This attribute exists on the Python module, but we failed to convert Python type: 'monai.networks.nets.milmodel.MilMode' to a TorchScript type.):
  File "/mnt/amproj/Code/MONAI/monai/networks/nets/milmodel.py", line 161
        sh = x.shape

        if self.mil_mode == MilMode.MEAN:


```

drbeh commented:
> I ran into another JIT issue, it fails the test on the older Pytorch version (1.7, over a year old), it looks like at that time Enum wasn't supported in JIT.
> 
> @wyli @Nic-Ma any suggestions? we can remove Enum, and use strings for MilMode, or we can skip this test ?
> 
> thanks
> 
> ```
> RuntimeError:
> Module 'MILModel' has no attribute 'mil_mode' (This attribute exists on the Python module, but we failed to convert Python type: 'monai.networks.nets.milmodel.MilMode' to a TorchScript type.):
>   File "/mnt/amproj/Code/MONAI/monai/networks/nets/milmodel.py", line 161
>         sh = x.shape
> 
>         if self.mil_mode == MilMode.MEAN:
> ```

@myron considering the tight deadline for the next release, let's use string instead of enum if it can fix the issue and move on.

myron commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,227 @@
+        self.mil_mode = mil_mode
+        print("MILModel with mode", mil_mode, "num_classes", num_classes)
+        self.attention = nn.Sequential()
+        self.transformer = None  # type: Optional[nn.Module]
```
this doesn't work for JIT. I get an error
`The TorchScript type system doesn't support instance-level annotations on empty non-base types in `__init__`. Instead, either 1) use a type annotation in the class body, or 2) wrap the type in `torch.jit.Attribute`.
`
and I don't do any annotations at all, JIT is fine, but mypy complains

myron commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,227 @@
+        mil_mode: MilMode = MilMode.ATT,
+        pretrained: bool = True,
+        backbone: Optional[Union[str, nn.Module]] = None,
+        backbone_nfeatures: Optional[int] = None,
```
his doesn't work for JIT. I get an error (see below)

myron commented on monai/networks/nets/milmodel.py:
```diff
@@ -0,0 +1,227 @@
+            nfc = net.fc.in_features  # save the number of final features
+            net.fc = torch.nn.Identity()  # remove final linear layer
+
+            self.extra_outputs = {}  # type: Dict[str, torch.Tensor]
```
this doesn't work for JIT. I get an error
The TorchScript type system doesn't support instance-level annotations on empty non-base types in init. Instead, either 1) use a type annotation in the class body, or 2) wrap the type in torch.jit.Attribute.
and I don't do any annotations at all, JIT is fine, but mypy complains

drbeh replied:
OK, I'll check on this later.

drbeh commented:
It looks good to me. Could you please take a look at the checks that are failing?

drbeh commented:
@Nic-Ma @wyli do you have any other commnet?

myron commented:
> @myron it looks good to me. Could you please take a look at the checks that are failing?

I don't know why it's failing. There is only 1 failed check, that seems to be unrelated to my PR

```
2021-11-14 16:01:06,684 - ERROR - check_hash failed 938869fb71b05f5ba345ab36dfbbc4388f53a550.
clang-format: 1.86MB [00:00, 2.41MB/s]                           
Download /__w/MONAI/MONAI/.clang-format-bin/clang-format failed: sha1 check of downloaded file failed: URL=https://oss-clang-format.s3.us-east-2.amazonaws.com/linux64/clang-format-linux64, filepath=/__w/MONAI/MONAI/.clang-format-bin/clang-format, expected sha1=9073602de1c4e1748f2feea5a0782417b20e3043.
Please remove /__w/MONAI/MONAI/.clang-format-bin/clang-format and retry.
```

Nic-Ma commented:
> > @myron it looks good to me. Could you please take a look at the checks that are failing?
> 
> I don't know why it's failing. There is only 1 failed check, that seems to be unrelated to my PR
> 
> ```
> 2021-11-14 16:01:06,684 - ERROR - check_hash failed 938869fb71b05f5ba345ab36dfbbc4388f53a550.
> clang-format: 1.86MB [00:00, 2.41MB/s]                           
> Download /__w/MONAI/MONAI/.clang-format-bin/clang-format failed: sha1 check of downloaded file failed: URL=https://oss-clang-format.s3.us-east-2.amazonaws.com/linux64/clang-format-linux64, filepath=/__w/MONAI/MONAI/.clang-format-bin/clang-format, expected sha1=9073602de1c4e1748f2feea5a0782417b20e3043.
> Please remove /__w/MONAI/MONAI/.clang-format-bin/clang-format and retry.
> ```

Hi @myron ,

I already submitted PR to fix it.

Thanks.

drbeh commented:
@Nic-Ma can you please run /build here?

yiheng-wang-nv reviewed (approved):
Looks good on me.

drbeh commented:
@Nic-Ma @wyli could you please run Blossom CI here?

Nic-Ma commented:
/build

Nic-Ma reviewed (approved):
Thanks for the great feature.
Looks good to me, I will slightly adjust the doc related things later.

Thanks.
